invenio_rest/dispatch.py

```diff
@@ -0,0 +1,199 @@
+        :param params: accept type's positional parameters.
+        :param kwparams: accept type's named parameters.
+        :param quality: accept type's quality (see HTTP ACCEPT HEADER RFC).
+        """
+        self.mime = mime
+        self.params = params
+        self.kwparams = kwparams
+        self.quality = quality
+
+    def __str__(self):
+        """Serialize as a valid accept type."""
+        result = self.mime
+        if self.params:
+            result += ';{}'.format(';'.join(self.params))
+        if self.kwparams:
+            result += ';{}'.format(';'.join(['{0}={1}'.format(k, v) for k, v in
+                                             self.kwparams.iteritems()]))
+        if self.quality:
+            result += ';q={}'.format(self.quality)
+        return result
+
+
+def parse_media_type(accept_type, quality):
+    """Parse the provided accept_type.
+
+    :param accept_type: accept type as returned by
+    py:func::`werkzeug.http.parse_accept_header`.
+    :param quality: accept type's quality.
+
+    :return: the parsed py:class::`AcceptType`.
+    """
+    elements = [elt.strip() for elt in accept_type.split(';')]
+    mime = elements[0]
+    elements = elements[1:]
+    params = []
+    kwparams = {}
+    for elt in elements:
+        if '=' in elt:
+            key, value = elt.split('=')
+            kwparams[key.strip()] = value.strip()
+        else:
+            params = elt
+    return AcceptType(mime, params, kwparams, quality)
+
+
+class HTTPAcceptDispatcherMiddleware(object):
+    """Application Dispatcher filtering on request's Accept Header."""
+
+    def __init__(self, dispatch_function=None,
+                 accept_header_rewriter=None, *args):
+        """Constructor.
+
+        :param dispatch_function: function used to dispatch the requests. It
+        will receive a list of py:class::`AcceptType` as argument.
+        :param accept_header_rewriter: function used to rewrite the accept.
+        header once an application has been chosen.
+        :param *args: additional arguments forwarded to the dispatch function.
+        """
+
+        self.dispatch_function = dispatch_function
+        self.accept_header_rewriter = accept_header_rewriter
+        self.args = args
+
+    def __call__(self, environ, start_response):
+        """Dispatch a request.
+
+        :param environ: request environment.
+        :param start_response: used to send the request's response.
+        """
+        accept_header = environ.get('HTTP_ACCEPT')
+        if accept_header is None:
+            app = self.dispatch_function(self.default_app,
+                                         *(self.args))
+            if self.accept_header_rewriter is not None:
+                environ['HTTP_ACCEPT'] = self.accept_header_rewriter([])
+            return app(environ, start_response)
+        else:
+            # parse all accept headers
+            parsed_accept = [parse_media_type(type, quality)
```

Comment:
I would probably go with ``type_`` or similar to not overwrite the python built-ins.